CCID: Multiple Low level and Class driver changes

kidbomb opened the pull request:
- Added initial support for Class Driver (supports IccPowerOn, IccPowerOff, GetSlotStatus and XfrBlock callbacks)
- Improved general documentation over CCID files
- Added support for abort control request
- Handling (and ignoring) other control requests
- Added initial Iso7816 library (currently only supports a versy simple ATR creation)
- Refactored some constants
- CCID_Task will use only buffer to send APDU responses
- Added JS host test application for CCID

kidbomb commented:
Hey @abcminiuser got a chance to look into this patch?

abcminiuser commented:
Not yet, I just got back from a two week business trip to the US a few hours ago - will give it a look once I'm over the 17 hour timeshift.

abcminiuser commented on Demos/Device/ClassDriver/CCID/CCID.c:
```diff
@@ -0,0 +1,223 @@
+uint8_t CALLBACK_CCID_XfrBlock(uint8_t slot, uint8_t *error, uint8_t * receivedBuffer, uint8_t receivedBufferSize, uint8_t * sendBuffer, uint8_t *sentBufferSize) {
+	if(slot < CCID_Interface.Config.TotalSlots) {
+		uint8_t okResponse[2] = {0x90, 0x00};
+		memcpy(sendBuffer, okResponse, sizeof(okResponse) * sizeof(uint8_t));
```
`sizeof(okResponse) * sizeof(uint8_t)` should be just `sizeof(okResponse)`, as `sizeof()` evaluated on a local array (not a pointer to an array) will include the stride length in the size calculation.

abcminiuser commented on Demos/Device/ClassDriver/CCID/CCID.h:
```diff
@@ -0,0 +1,83 @@
+	/* Callbacks */
+		uint8_t CALLBACK_CCID_IccPowerOff(uint8_t slot, uint8_t *error);
+		uint8_t CALLBACK_CCID_GetSlotStatus(uint8_t slot, uint8_t *error);
+
```
Missing callback prototypes here, such as `CALLBACK_CCID_Abort()`.

abcminiuser commented on Demos/Device/ClassDriver/CCID/Descriptors.c:
```diff
@@ -0,0 +1,219 @@
+			.MaxSlotIndex			= 0x00,
+			.VoltageSupport			= CCID_VOLTAGESUPPORT_5V,
+			.Protocols				= CCID_PROTOCOLS_T1,
+			.DefaultClock			= 16000, //16MHz
```
Might be worth a `CCID_DESCRIPTOR_CLOCK(hz)` or similar macro in the library, which does the /1000 scaling for the user.

abcminiuser commented on Demos/Device/ClassDriver/CCID/Descriptors.c:
```diff
@@ -0,0 +1,219 @@
+ *  form, and is read out upon request by the host when the appropriate string ID is requested, listed in the Device
+ *  Descriptor.
+ */
+const USB_Descriptor_String_t PROGMEM ManufacturerString = USB_STRING_DESCRIPTOR(L"CCID Manufacturer");
```
I think I'll change all the existing demo manufacturer descriptors to "LUFA Library" rather than my own name, can you also change that here please?

abcminiuser commented on Demos/Device/ClassDriver/CCID/Descriptors.h:
```diff
@@ -0,0 +1,97 @@
+			USB_CCID_Descriptor_t					CCID_SmartCard;
+			USB_Descriptor_Endpoint_t				CCID_BulkInEndpoint;
+			USB_Descriptor_Endpoint_t				CCID_BulkOutEndpoint;
+
```
Nit: remove blank line here.

abcminiuser commented on Demos/Device/ClassDriver/CCID/Lib/Iso7816.c:
```diff
@@ -0,0 +1,46 @@
+
+	uint8_t interfaceBytesPresence = 0;
+	
+	uint8_t historycalBytes[14] = "Lufa CCID Demo"; //must be equal or less than 15
```
Nit: "LUFA", not "Lufa" as it is an acronym.

abcminiuser replied:
Spelling: `historycalBytes` -> `historicalBytes`?

abcminiuser commented on Demos/Device/LowLevel/CCID/CCID.txt:
```diff
@@ -0,0 +1,54 @@
+ *
+ *  The following list indicates what microcontrollers were testes with this demo.
+ *
+ *  \li ATMEGA16U2
```
Should expand this to other supported devices (looks like everything except the 8U2 if it requires a lot of USB DPRAM or regular RAM. Follow the template from the other demo text file, please.

abcminiuser commented on LUFA/Drivers/USB/Class/Common/CCIDClassCommon.h:
```diff
@@ -91,34 +105,44 @@
 		};
 
-		/** Enum for the CCID class specific control requests that can be issued by the USB bus host. */
+		/** Enum for the Mass Storage class specific control requests that can be issued by the USB bus host. */
```
CCID, not Mass Storage.

kidbomb commented:
Just rebased/squashed my commits with my code review changes.

abcminiuser commented:
Looking decent now, I just have to find time to reformat it and fully integrate it. Sorry, I know it's been a slow process - me being out of the country for two weeks hasn't helped, and right now a bunch of my stuff is in boxes for the impending home move.

I'll do this as soon as I can, hopefully won't be too much longer. If I have time this week after work I'll give it a go.

kidbomb commented:
Hey @abcminiuser got a chance to take a look into this already?

abcminiuser commented:
A rather long afternoon of merging, squashing, rebasing, reformatting and more and I've got a minimal `feature-ccid` branch now with all the changes. I think this is pretty much ready to go, minus some small last changes I need from you:

1) I need descriptions for the callbacks in `LUFA\Drivers\USB\Class\Device\CCIDClassDevice.h` that aren't currently documented. Can you please add these, including the return codes and parameter descriptions (see the functions above it for examples.

2) The `CALLBACK_CCID_XfrBlock` callback doesn't seem to be used, and there's a `TODO:` in the handing for that request in `LUFA\Drivers\USB\Class\Device\CCIDClassDevice.c`. Should the callback be called there?

3) Same deal with the low level device demo, there's a `TODO:` still in the XfrBlock message handling.

Once those are sorted this is ready for merge. Since this is a squashed branch I recommend you move your existing branch to another name, and diff it against the upstream feature branch to make sure I didn't mess anything up or miss any of your changes.

kidbomb commented:
Just checked it now - I added the requested documentation and removed the XfrBlock callbacks for now. As soon as I get the right design for it I can create a review for it separately.

abcminiuser commented:
Added const qualifiers, context parameters to the CCID device driver callbacks and some additional reformatting to match existing code style.
